### ISSUE TITLE
Sonoff POWR316 does not report Energy entity - reported as "unknown"

### ISSUE THREAD

**Issue by kformosa** [AlexxIT/SonoffLAN#1129]
Sonoff POWR316 Firmware: 1.0.7
HA: 2023.3.1

In the integration:
-  current
- power 
- voltage
are all being reported correctly

Energy being reported as "unknown". Tried switching being "auto", "local" and "cloud" but still nothing.
I prefer "local" mode since I have many other Sonoff devices which work flawlessly.

As a workaround I implemented as you suggested an Integration Sensor in HA, which is doing the job well but doesn't feel clean.

Thanks!

**Comment by lbouriez** [comment]
@kformosa can you share the sensor energy template please ?

**Comment by calas80** [comment]
Some Problem.
Can you tell me the solution you are talking about?

**Comment by HomeAssistantProTeam** [comment]
> 



> Show device diagnostics

{
  "home_assistant": {
    "installation_type": "Home Assistant OS",
    "version": "2023.3.6",
    "dev": false,
    "hassio": true,
    "virtualenv": false,
    "python_version": "3.10.10",
    "docker": true,
    "arch": "x86_64",
    "timezone": "Europe/Rome",
    "os_name": "Linux",
    "os_version": "5.15.90",
    "supervisor": "2023.03.3",
    "host_os": "Home Assistant OS 9.5",
    "docker_version": "20.10.22",
    "chassis": "vm",
    "run_as_root": true
  },
  "custom_components": {
    "sonoff": {
      "version": "3.4.0",
      "requirements": [
        "pycryptodome>=3.6.6"
      ]
    },
    "hacs": {
      "version": "1.31.0",
      "requirements": [
        "aiogithubapi>=22.10.1"
      ]
    }
  },
  "integration_manifest": {
    "domain": "sonoff",
    "name": "Sonoff",
    "config_flow": true,
    "documentation": "https://github.com/AlexxIT/SonoffLAN",
    "issue_tracker": "https://github.com/AlexxIT/SonoffLAN/issues",
    "codeowners": [
      "@AlexxIT"
    ],
    "dependencies": [
      "http",
      "zeroconf"
    ],
    "requirements": [
      "pycryptodome>=3.6.6"
    ],
    "version": "3.4.0",
    "iot_class": "local_push",
    "is_built_in": false
  },
  "data": {
    "version": "5406fa7",
    "cloud_auth": true,
    "config": null,
    "options": {
      "mode": "auto",
      "debug": false
    },
    "errors": [
      {
        "name": "custom_components.sonoff.core.ewelink.local",
        "message": [
          "100186d0e7 => Local4 | {'switches': [{'outlet': 0, 'switch': 'off'}]} <= {'sequence': '1680163365000', 'seq': 53, 'error': 400, 'encrypt': True}",
          "100186d0e7 => Local4 | {'switches': [{'outlet': 0, 'switch': 'on'}]} <= {'sequence': '1680163366001', 'seq': 55, 'error': 400, 'encrypt': True}"
        ],
        "level": "WARNING",
        "source": [
          "custom_components/sonoff/core/ewelink/local.py",
          214
        ],
        "timestamp": 1680163366.8364754,
        "exception": "",
        "count": 2,
        "first_occurred": 1680163366.0239687
      },
      {
        "name": "custom_components.sonoff.core.ewelink.cloud",
        "message": [
          "Cloud ERROR: {'error': 504, 'reason': 'Request Timeout', 'deviceid': '100186d0e7', 'apikey': '35d06ceb-c8dc-4b5c-8ba0-47569b3dbb0b', 'sequence': '1680158789000'}",
          "Cloud ERROR: {'error': 504, 'reason': 'Request Timeout', 'deviceid': '100186d0e7', 'apikey': '35d06ceb-c8dc-4b5c-8ba0-47569b3dbb0b', 'sequence': '1680162390000'}",
          "Cloud ERROR: {'error': 504, 'reason': 'Request Timeout', 'deviceid': '100186d0e7', 'apikey': '35d06ceb-c8dc-4b5c-8ba0-47569b3dbb0b', 'sequence': '1680165990000'}",
          "Cloud ERROR: {'error': 504, 'reason': 'Request Timeout', 'deviceid': '100186d0e7', 'apikey': '35d06ceb-c8dc-4b5c-8ba0-47569b3dbb0b', 'sequence': '1680170370000'}",
          "Cloud ERROR: {'error': 504, 'reason': 'Request Timeout', 'deviceid': '100186d0e7', 'apikey': '35d06ceb-c8dc-4b5c-8ba0-47569b3dbb0b', 'sequence': '1680174841000'}"
        ],
        "level": "WARNING",
        "source": [
          "custom_components/sonoff/core/ewelink/cloud.py",
          342
        ],
        "timestamp": 1680174846.0789883,
        "exception": "",
        "count": 19,
        "first_occurred": 1680108391.2313602
      }
    ],
    "devices": {
      "100186d0e7": {
        "uiid": 190,
        "params": {
          "bindInfos": "***",
          "version": 8,
          "demNextFetchTime": 1680127200000,
          "ssid": "***",
          "bssid": "***",
          "fwVersion": "1.0.2",
          "switches": [
            {
              "switch": "off",
              "outlet": 0
            }
          ],
          "configure": [
            {
              "startup": "off",
              "outlet": 0
            }
          ],
          "pulses": [
            {
              "pulse": "off",
              "switch": "off",
              "outlet": 0,
              "width": 500
            }
          ],
          "rssi": -59,
          "threshold": {
            "actPow": {
              "min": 10,
              "max": 400000
            },
            "voltage": {
              "min": 18500,
              "max": 26400
            },
            "current": {
              "min": 10,
              "max": 1600
            }
          },
          "current": 0,
          "voltage": 0,
          "power": 0,
          "uiActive": 180,
          "timeZone": 2,
          "dayKwh": 0,
          "monthKwh": 0
        },
        "model": "POWR316",
        "online": true,
        "localtype": "plug",
        "host": null
      },
      "100186cf65": {
        "uiid": 190,
        "params": {
          "bindInfos": "***",
          "version": 8,
          "demNextFetchTime": 1680127200000,
          "ssid": "***",
          "bssid": "***",
          "fwVersion": "1.0.7",
          "switches": [
            {
              "outlet": 0,
              "switch": "off"
            }
          ],
          "configure": [
            {
              "startup": "off",
              "outlet": 0
            }
          ],
          "pulses": [
            {
              "pulse": "off",
              "switch": "off",
              "outlet": 0,
              "width": 500
            }
          ],
          "sledOnline": "on",
          "rssi": -58,
          "current": 16,
          "voltage": 23825,
          "power": 3181,
          "timeZone": 2,
          "dayKwh": 2,
          "monthKwh": 2,
          "uiActive": 180,
          "getHoursKwh": {
            "start": 0,
            "end": 743
          },
          "threshold": {
            "actPow": {
              "min": 10,
              "max": 400000
            },
            "voltage": {
              "min": 18500,
              "max": 26400
            },
            "current": {
              "min": 10,
              "max": 1600
            }
          }
        },
        "model": "POWR316",
        "online": false,
        "localtype": null,
        "host": null
      }
    }
  }
}

**Comment by AlexxIT** [comment]
Looks like `'error': 504` the reason. Energy can't be loaded from cloud. I'll check it with my devices

**Comment by AlexxIT** [comment]
Can you share this device to my account? sonofflan@gmail.com
I don't like `'error': 400`. Something goes wrong for LAN command

**Comment by HomeAssistantProTeam** [comment]
> Can you share this device to my account? [sonofflan@gmail.com](mailto:sonofflan@gmail.com) I don't like `'error': 400`. Something goes wrong for LAN command

Done

**Comment by AlexxIT** [comment]
@HomeAssistantProTeam Thanks! One of your devices offline. Second with empty energy history.
I need some history for implementation :)

**Comment by HomeAssistantProTeam** [comment]
> @HomeAssistantProTeam Thanks! One of your devices offline. Second with empty energy history. I need some history for implementation :)

You're right, it's a new device. I connect a running pc

**Comment by AlexxIT** [comment]
What about offline device? Can you turn it on?

**Comment by AlexxIT** [comment]
It is still offline for me: S1 - offline, S2 - online with empty energy history.

**Comment by HomeAssistantProTeam** [comment]
> It is still offline for me: S1 - offline, S2 - online with empty energy history.

S2 is the only device I can keep online at the moment. S2 has been on for a few hours with a pc connected. Unfortunately I have no history to give you, they are new devices that I bought now. I will leave S2 on with pc plugged in for as long as you need

**Comment by kformosa** [comment]
@AlexxIT I shared my POWR316 device with you. Please let me know if you need something else

**Comment by kformosa** [comment]
How can I test it? From HACS, there is only an update to v3.5.0 and I can see that this commit is after this version has been built.
Should I wait until 3.6 or another minor release is done?

**Comment by AlexxIT** [comment]
You can install it via HACS reinstall feature

**Comment by calas80** [comment]
> You can install it via HACS reinstall feature

HI. I did the update 3.5.0 but still see unknown. What am I doing wrong?

{
  "home_assistant": {
    "installation_type": "Home Assistant OS",
    "version": "2023.3.6",
    "dev": false,
    "hassio": true,
    "virtualenv": false,
    "python_version": "3.10.10",
    "docker": true,
    "arch": "x86_64",
    "timezone": "Europe/Rome",
    "os_name": "Linux",
    "os_version": "5.15.90",
    "supervisor": "2023.03.3",
    "host_os": "Home Assistant OS 9.5",
    "docker_version": "20.10.22",
    "chassis": "vm",
    "run_as_root": true
  },
  "custom_components": {
    "sonoff": {
      "version": "3.5.0",
      "requirements": [
        "pycryptodome>=3.6.6"
      ]
    },
    "hacs": {
      "version": "1.31.0",
      "requirements": [
        "aiogithubapi>=22.10.1"
      ]
    }
  },
  "integration_manifest": {
    "domain": "sonoff",
    "name": "Sonoff",
    "config_flow": true,
    "documentation": "https://github.com/AlexxIT/SonoffLAN",
    "issue_tracker": "https://github.com/AlexxIT/SonoffLAN/issues",
    "codeowners": [
      "@AlexxIT"
    ],
    "dependencies": [
      "http",
      "zeroconf"
    ],
    "requirements": [
      "pycryptodome>=3.6.6"
    ],
    "version": "3.5.0",
    "iot_class": "local_push",
    "is_built_in": false
  },
  "data": {
    "version": "6935784",
    "cloud_auth": true,
    "config": null,
    "options": {
      "mode": "auto",
      "debug": false
    },
    "errors": [
      {
        "name": "custom_components.sonoff.core.ewelink.cloud",
        "message": [
          "Cloud ERROR: {'error': 504, 'reason': 'Request Timeout', 'deviceid': '100186d0e7', 'apikey': '35d06ceb-c8dc-4b5c-8ba0-47569b3dbb0b', 'sequence': '1680250756000'}"
        ],
        "level": "WARNING",
        "source": [
          "custom_components/sonoff/core/ewelink/cloud.py",
          342
        ],
        "timestamp": 1680250761.0800192,
        "exception": "",
        "count": 1,
        "first_occurred": 1680250761.0800192
      }
    ],
    "devices": {
      "100186d0e7": {
        "uiid": 190,
        "params": {
          "bindInfos": "***",
          "version": 8,
          "demNextFetchTime": 1680300000000,
          "ssid": "***",
          "bssid": "***",
          "fwVersion": "1.0.7",
          "switches": [
            {
              "outlet": 0,
              "switch": "on"
            }
          ],
          "configure": [
            {
              "startup": "off",
              "outlet": 0
            }
          ],
          "pulses": [
            {
              "pulse": "off",
              "switch": "off",
              "outlet": 0,
              "width": 500
            }
          ],
          "rssi": -62,
          "threshold": {
            "actPow": {
              "min": 10,
              "max": 400000
            },
            "voltage": {
              "min": 18500,
              "max": 26400
            },
            "current": {
              "min": 10,
              "max": 1600
            }
          },
          "current": 17,
          "voltage": 23565,
          "power": 3241,
          "uiActive": 60,
          "timeZone": 2,
          "dayKwh": 33,
          "monthKwh": 61,
          "getHoursKwh": {
            "start": 0,
            "end": 743
          },
          "only_device": {
            "ota": "success",
            "ota_fail_reason": 0
          }
        },
        "model": "POWR316",
        "online": true,
        "local": true,
        "localtype": "plug",
        "host": "192.168.8.224:8081"
      },
      "100186cf65": {
        "uiid": 190,
        "params": {
          "bindInfos": "***",
          "version": 8,
          "demNextFetchTime": 1680127200000,
          "ssid": "***",
          "bssid": "***",
          "fwVersion": "1.0.7",
          "switches": [
            {
              "outlet": 0,
              "switch": "off"
            }
          ],
          "configure": [
            {
              "startup": "off",
              "outlet": 0
            }
          ],
          "pulses": [
            {
              "pulse": "off",
              "switch": "off",
              "outlet": 0,
              "width": 500
            }
          ],
          "sledOnline": "on",
          "rssi": -58,
          "current": 16,
          "voltage": 23825,
          "power": 3181,
          "timeZone": 2,
          "dayKwh": 2,
          "monthKwh": 2,
          "uiActive": 180,
          "getHoursKwh": {
            "start": 0,
            "end": 743
          },
          "threshold": {
            "actPow": {
              "min": 10,
              "max": 400000
            },
            "voltage": {
              "min": 18500,
              "max": 26400
            },
            "current": {
              "min": 10,
              "max": 1600
            }
          }
        },
        "model": "POWR316",
        "online": false,
        "local": null,
        "localtype": null,
        "host": null
      }
    }
  }
}

**Comment by HomeAssistantProTeam** [comment]
> @calas80 you did not install the master version

I downloaded the SonoffLAN-master.zip file from here: https://github.com/AlexxIT/SonoffLAN.
I took all contents of **\SonoffLAN-master\custom_components\sonoff folder** and copied it to **\config\custom_components\sonoff**.
reboot.
it's correct?

**Comment by AlexxIT** [comment]
It is easier to do it through HACS, but your way also works

**Comment by calas80** [comment]
I have installed the master version from HACS.
I rebooted
I removed the integration and downloaded the master again
restarted
but I always see unknown

![image](https://user-images.githubusercontent.com/48358142/229103826-8c886b40-c0b9-49cc-87d8-73e19df34ac2.png)

**Comment by AlexxIT** [comment]
https://github.com/AlexxIT/SonoffLAN/releases/tag/v3.5.1

**Comment by calas80** [comment]
> https://github.com/AlexxIT/SonoffLAN/releases/tag/v3.5.1

I'm so sorry! I downloaded 3.5.1, rebooted and reintegrated but it's still unknown! I have the Sonoff POWR316. Am I the problem?

**Comment by AlexxIT** [comment]
@calas80 you need to collect debug logs

**Comment by lbouriez** [comment]
I can confirm that on the master branch it's now working
![image](https://user-images.githubusercontent.com/4749766/229118994-ea98d108-fe21-4986-a417-ab9083425111.png)

**Comment by kformosa** [comment]
Thanks @AlexxIT - I updated the integration and now I have Daily and Monthly Energy entities.

![image](https://user-images.githubusercontent.com/37046796/229291690-9f4822cb-578f-47ab-8052-9716526ffa2d.png)

But adding the Energy Daily sensor to the "Energy Dashboard", I am getting "Last reset missing". It seems that you need to add this - I couldn't

![image](https://user-images.githubusercontent.com/37046796/229291783-df4f2519-0340-4787-8407-6e2157fd4254.png)

**Comment by lbouriez** [comment]
Wait a little bit, i had the same and now it works

**Comment by lbouriez** [comment]
![Screenshot_20230401-093739](https://user-images.githubusercontent.com/4749766/229292369-45322d40-82c6-4e65-a765-ad6c27d42bd8.png)
I am on my phone... So a screenshot will do.
Add this template and use that in the Energy dashboard

**Comment by kformosa** [comment]
Seems to be working - at least added in to the Energy dashboard without warning.

The gotcha for me was the new and legacy formats for the template - wasted 1 hour figuring what was the error!

Will check in few hours to see if all is good in the Energy dashboard

**Comment by AlexxIT** [comment]
@lbouriez are you sure daily sensor should be "total increasing"?

**Comment by kformosa** [comment]
Yes @AlexxIT - it has to be "total increasing" to display the usage properly in the "Energy" dashboard.

It expects that the energy is incrementing daily - keeps track of the last energy value for the next day. Otherwise it will display 0 for the following day.

Mine was first set up as "measurement" and had this problem. Changed it to "total increasing" and issue resolved.

**Comment by AlexxIT** [comment]
I thought "total increasing" means that the value will always increasing. And it doesn't drops to zero every day.

**Comment by kformosa** [comment]
It doesn't drop to zero every day:

So if I use 2kWh every day, after 5 days, the sensor with total_increasing will have a value of 10kWh.
And that sensor in the Energy dashboard will mark usage of 2kWh per day. 
Energy dashboard needs a sensor which accumulates usage and will apportion the value per day (knowing the total value from the previous day)

At least this is my understanding from trying different things.

**Comment by kformosa** [comment]
@AlexxIT 
I'm seeing WARNING logs in my HA:

2```
023-04-03 13:35:42.973 WARNING (MainThread) [custom_components.sonoff.core.ewelink.cloud] Cloud ERROR: {'error': 504, 'reason': 'Request Timeout', 'deviceid': '100186d41d', 'apikey': 'xxxxx', 'sequence': '1680521737000'}
2023-04-03 13:35:47.998 WARNING (MainThread) [homeassistant.helpers.entity] Update of sensor.sonoff_100186d41d_energy is taking over 10 seconds
2023-04-03 13:43:48.013 WARNING (MainThread) [homeassistant.helpers.entity] Update of sensor.sonoff_100186d41d_energy is taking over 10 seconds
2023-04-03 13:43:48.462 WARNING (MainThread) [custom_components.sonoff.core.ewelink.cloud] Cloud ERROR: {'error': 504, 'reason': 'Request Timeout', 'deviceid': '100186d41d', 'apikey': 'xxxxx', 'sequence': '1680522218000'}
2023-04-03 15:21:13.175 WARNING (MainThread) [custom_components.sonoff.core.ewelink.cloud] Cloud ERROR: {'error': 504, 'reason': 'Request Timeout', 'deviceid': '100186d41d', 'apikey': 'xxxxxx', 'sequence': '1680528068000'}
2023-04-03 15:26:13.182 WARNING (MainThread) [custom_components.sonoff.core.ewelink.cloud] Cloud ERROR: {'error': 504, 'reason': 'Request Timeout', 'deviceid': '100186d41d', 'apikey': 'xxxxxx', 'sequence': '1680528368000'}
2023-04-03 15:27:18.190 WARNING (MainThread) [homeassistant.helpers.entity] Update of sensor.sonoff_100186d41d_energy is taking over 10 seconds
2023-04-03 15:27:18.471 WARNING (MainThread) [custom_components.sonoff.core.ewelink.cloud] Cloud ERROR: {'error': 504, 'reason': 'Request Timeout', 'deviceid': '100186d41d', 'apikey': 'xxxxx', 'sequence': '1680528428001'}
```

Can it be the integration rate limits the update to say 1 every 2 or 5 minutes? 
Maybe Sonoff doesn't not accept this many requests. I don't know

**Comment by AlexxIT** [comment]
I don't think it helps. Check timestamps. Energy just a heavy request for big history.

**Comment by kformosa** [comment]
I am too experiencing my POWR316 switching OFF after approx. 1 hour. This device should always be ON since I mainly use it to log power usage. On the eWeLink App I selected Power-on state ON but it doesn't make any effect. 

It used to work fine until 2 days ago - random switching OFF just happened. I also noticed that the history on the eWeLink App is all 0; i.e. it seems like not being recorded or deleted. Only have history since around 1-2 April when I think @AlexxIT pushed an update. Again, I don't know if it is related or not.

I can see a number of ERRORS in the log.

I had to remove the device from HA since I cannot afford this random switching OFF. I will test my device being connected to power but without a config on HA and see if the random switching OFF still happens or not.
